### ISSUE TITLE
There is no simple way to prevent the truncated JPEG loading

### ISSUE THREAD

**Issue by bozaro** [libvips/libvips#2309]
# Original issue
I want to deny uploading truncated JPEG to microservice: this error that can silenly creep in on different layers on the way from the user to the microservice.

The original idea was to use the `fail` flag on JPEG loading. This flag treats all warnings as errors. As a result, I get an error while loading a truncated JPEG, as I am trying.

Unfortunately, some warnings appear in pictures that do not strictly comply with the JPEG standard.
For example, some Samsung Android devices like Samsung Galaxy J1 SM-J120F make JPEG images with `Invalid SOS parameters for sequential JPEG` warning. These pictures are processed by many libraries without any complaints.

These pictures also stop loading with `fail` flag.

The situation is especially complicated by the fact that when paraller image processing, there is no rellation between warning message and image.

# My workaround
I have not found an easy way to solve this problem and use this logic:
 - check image format by file signature;
 - for JPEG make double check with directly `libjpeg` library calls with ignoring all warnings except JWRN_JPEG_EOF;
 - load image for processing with `libvips` with `fail == FALSE` flag.

# Possible solutions
I don't see good simple solution.
I see the following options:
 - ignore some errors inside `libvips` - this might solve my problem, but I don't think `libvips` should be doing this: the warning severity is subjective;
 - add some warnings to error handler - move the decision about the problem severity to the developer side.

**Comment by jcupitt** [comment]
Hi @bozaro,

Yes, it's a tricky issue. Some truncations will not cause errors or warnings, for example (try removing the final few bytes of a JPEG). Every load library takes a slightly different view of what counts as a warning and what is an error, so it's especially hard for libvips to present them in a uniform way.

You probably saw the bit of code in the libvips JPEG loader:

https://github.com/libvips/libvips/blob/master/libvips/foreign/jpeg2vips.c#L252

Perhaps we need a new loader flag (`fail-on-truncated`?) just for this case, since truncated files are so common. It could default to the value of `fail`, for compatibility. Does that sound reasonable?

**Comment by bozaro** [comment]
It seems to me appropriate would be two mechanisms complementary to each other:
 - warning callback for move the decision about the problem severity to the developer side;
 - `fail-on-truncated` for simple prevent truncated file shortcut.

But I am not using `libvips` directly and I can't imagine how well it fits into his internal structure.

**Comment by jcupitt** [comment]
callbacks are awkward for many language bindings. Let's add `fail-on-truncated` instead. All the loaders will need a small change to support this, unfortunately.

**Comment by jcupitt** [comment]
I've started a PR to add this feature, any comments or testing welcome etc.

**Comment by jcupitt** [comment]
The proposed API has shifted a but, it's now:

```python
image = pyvips.Image.new_from_file("xxx.jpg", fail_on="truncated")
```

ie. only fail on truncations, everything else is just a warning.

**Comment by jcupitt** [comment]
This is now merged and will be a feature of 8.12. Thanks for the idea!
